### PR TITLE
📦 NEW: Add HTTP Rate Limiting

### PR DESCRIPTION
**Pull request by aymanizz**

### CODE REVIEW

File: smalld/smalld.py [hunk 1]
```diff
@@ -204,10 +206,15 @@ def send_request(self, method, path, payload="", attachments=None):
             args = {"json": payload}
         else:
             args = {}
+        req = requests.Request(method, f"{self.base_url}/{path}", **args)
```
Comment (ianagbip1oti) [inline smalld/smalld.py:209]: So, we have the method and the url, and we pack them into a request so we can pass it to on_request, where we unpack it into method and url?

File: test/test_ratelimit.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+
+@pytest.mark.parametrize(
+    "url, group",
+    [
+        ("channels/2909267986263572999", "channels/2909267986263572999"),
+        ("guilds/197038439483310086", "guilds/197038439483310086"),
+        ("webhooks/223704706495545344", "webhooks/223704706495545344"),
+        ("/channels/2909267986263572999/", "channels/2909267986263572999"),
+        ("/guilds/197038439483310086/members/{user.id}", "guilds/197038439483310086"),
```
Comment (ianagbip1oti) [inline test/test_ratelimit.py:148]: Am I understanding that any url starting with `/guilds/197038439483310086` would be mapped to the same bucket?

File: smalld/ratelimit.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+        bucket_id = url_group(url) or bucket_id
+        if not bucket_id:
+            return self.no_ratelimit_bucket
+
+        try:
+            bucket = self.buckets[bucket_id]
+        except KeyError:
+            bucket = self.buckets[bucket_id] = ResourceRateLimitBucket(bucket_id)
+        self.resource_buckets[key] = bucket
```
Comment (ianagbip1oti) [inline smalld/ratelimit.py:106]: We add an entry here for every url we might hit and never remove it?

File: smalld/resources/ratelimit_buckets [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Audit Log
```
Comment (ianagbip1oti) [inline smalld/resources/ratelimit_buckets:1]: So as well as updating some \d to \w the \\ also had to be removed. Hmm.

It is actually cleaner without the \\. Will just have to write a customer parser for Java too (it's not a difficult format to parse tho)

File: smalld/smalld.py [hunk 5]
```diff
@@ -205,9 +207,11 @@ def send_request(self, method, path, payload="", attachments=None):
         else:
             args = {}
 
-        r = self.session.request(method, f"{self.base_url}/{path}", **args)
+        self.limiter.on_request(method, path)
+        res = self.session.request(method, f"{self.base_url}/{path}", **args)
```
Comment (ianagbip1oti) [inline smalld/smalld.py:211]: What does requests to on non-200 responses? Does it throw an exception or just give you the response?

Comment (aymanizz) [reply smalld/smalld.py:211]: It returns the response, for raising an exception you just call `res.raise_for_status()`.

**Pull request merged**